packages/valory/skills/mech_interact_abci/behaviours/purchase_subcription.py

```diff
@@ -0,0 +1,696 @@
+            contract_callable="build_approval_tx",
+            data_key="data",
+            placeholder="_subscription_token_approval_tx_data",
+            spender=self.params.lock_payment_condition_address,
+            amount=SUBSCRIPTION_COST,
+            chain_id=self.params.mech_chain_id,
+        )
+        if not status:
+            self.context.logger.error("Failed to build data for a USDC approval tx.")
+            return False
+
+        batch = MultisendBatch(
+            to=self.nvm_config.subscription_token_address,
+            data=self.subscription_token_approval_tx_data,
+        )
+        self.multisend_batches.append(batch)
+        self.context.logger.info("Built transaction to approve USDC spending.")
+        return True
+
+    def _build_create_fulfill_tx_data(self):
+        self.context.logger.info(
+            f"Creating a fulfill tx with {self.fulfill_for_delegate_params=} and {self.fulfill_params=}."
+        )
+        status = self.contract_interact(
+            performative=ContractApiMessage.Performative.GET_RAW_TRANSACTION,  # type: ignore
+            contract_address=self.nvm_config.subscription_provider_address,
+            contract_public_id=SubscriptionProvider.contract_id,
+            contract_callable="build_create_fulfill_tx",
+            data_key="data",
+            placeholder="_create_fulfill_tx_data",
+            agreement_id_seed=self.agreement_id_seed,
+            did=self.nvm_config.did,
+            fulfill_for_delegate_params=self.fulfill_for_delegate_params,
+            fulfill_params=self.fulfill_params,
+            chain_id=self.params.mech_chain_id,
+        )
+        if not status:
+            self.context.logger.error("Failed to build data for a fulfill tx.")
+            return False
+
+        batch = MultisendBatch(
+            to=self.nvm_config.subscription_provider_address,
+            data=self.fulfill_tx_data,
+        )
+        self.multisend_batches.append(batch)
+        self.context.logger.info("Built transaction to fulfill.")
+        return True
+
+    def _get_approval_steps(self) -> List[WaitableConditionType]:
+        """Get the approval steps, if necessary, otherwise return an empty list."""
+        return (
+            [self._build_subscription_token_approval_tx_data]
+            if ChainType(self.params.mech_chain_id) == ChainType.BASE
+            else []
+        )
+
+    def _prepare_safe_tx(self) -> Generator:
+        """Prepare a multisend safe tx for buying an NVM subscription."""
+        steps = [
+            self._get_ddo_register,
+            self._get_ddo_data,
+            self._get_agreement_id,
+            self._get_lock_hash,
+            self._get_lock_id,
+            self._get_transfer_nft_hash,
+            self._get_transfer_id,
+            self._get_escrow_payment_hash,
+            self._get_escrow_id,
+            self._build_create_agreement_tx_data,
+            *self._get_approval_steps(),
```

Comment:
Steps have to be swapped, approval needs to happen before create agreeement

Reply:
d7297e03a6a6eee248facca02a08196baa41afc4